lib/Dialect/Arc/Transforms/LatencyRetiming.cpp

```diff
@@ -0,0 +1,136 @@
+#include "circt/Support/SymCache.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "arc-latency-retiming"
+
+using namespace circt;
+using namespace arc;
+
+//===----------------------------------------------------------------------===//
+// Patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LatencyRetimingStatistics {
+  unsigned numOpsRemoved = 0;
+  unsigned latencyUnitsSaved = 0;
+};
+
+/// Absorb the latencies from predecessor states to collapse shift registers and
+/// reduce the overall amount of latency units in the design.
+struct LatencyRetimingPattern : OpRewritePattern<StateOp> {
+  LatencyRetimingPattern(MLIRContext *context, SymbolCache &symCache,
+                         LatencyRetimingStatistics &statistics)
+      : OpRewritePattern<StateOp>(context), symCache(symCache),
+        statistics(statistics) {}
+
+  LogicalResult matchAndRewrite(StateOp op,
+                                PatternRewriter &rewriter) const final;
+
+private:
+  SymbolCache &symCache;
+  LatencyRetimingStatistics &statistics;
+};
+
+} // namespace
+
+LogicalResult
+LatencyRetimingPattern::matchAndRewrite(StateOp op,
+                                        PatternRewriter &rewriter) const {
+  unsigned minPrevLatency = UINT_MAX;
+  SetVector<StateOp> predecessors;
+
+  for (auto input : op.getInputs()) {
+    auto predState = input.getDefiningOp<StateOp>();
+    if (!predState)
+      return failure();
+
+    if (predState->hasAttr("name") || predState->hasAttr("names"))
+      return failure();
+
+    if (predState == op)
+      return failure();
+
+    if (predState.getLatency() != 0 && op.getLatency() != 0 &&
+        predState.getClock() != op.getClock())
+      return failure();
+
+    if (predState.getEnable() || predState.getReset() || op.getReset() ||
+        op.getEnable())
+      return failure();
+
+    if (llvm::any_of(predState->getUsers(),
+                     [&](auto *user) { return user != op; }))
+      return failure();
+
+    predecessors.insert(predState);
+    minPrevLatency = std::min(minPrevLatency, predState.getLatency());
+  }
+
+  if (minPrevLatency == 0 || minPrevLatency == UINT_MAX)
+    return failure();
+
+  op.setLatency(op.getLatency() + minPrevLatency);
+  for (auto prevStateOp : predecessors) {
+    if (!op.getClock() && !op->getParentOfType<ClockDomainOp>())
+      op.getClockMutable().assign(prevStateOp.getClock());
+
+    statistics.latencyUnitsSaved += minPrevLatency;
+    auto newLatency = prevStateOp.getLatency() - minPrevLatency;
+    prevStateOp.setLatency(newLatency);
+
+    if (newLatency > 0)
+      continue;
+
+    prevStateOp.getClockMutable().clear();
+    if (cast<DefineOp>(symCache.getDefinition(prevStateOp.getArcAttr()))
+            .isPassthrough()) {
+      rewriter.replaceOp(prevStateOp, prevStateOp.getInputs());
+      ++statistics.numOpsRemoved;
+    }
```

Comment:
Clever to check for passthroughs that have become trivial :+1:

Reply:
Unfortunately, that's a simplification that we cannot put in the state op canonicalizer because we don't have a symbol cache there.